[GStreamer][WebRTC] Excessive tag events with bitrate update at startup

cadubentzen opened the pull request:
#### 18068fac3198661784deaaa7c4221666999a894f
<pre>
[GStreamer][WebRTC] Excessive tag events with bitrate update at startup
<a href="https://bugs.webkit.org/show_bug.cgi?id=290045">https://bugs.webkit.org/show_bug.cgi?id=290045</a>

Reviewed by Philippe Normand.

Codec parsers like h264parse emit too many tag events at startup, while carrying
only small bitrate updates. As time passes, bitrate updates are sent
less frequently.

Reduce the overhead in the pipelines at startup by forwarding tag events only every
1 second in RealtimeIncomingSourceGStreamer.

Canonical link: <a href="https://commits.webkit.org/292399@main">https://commits.webkit.org/292399@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37acfc65f290db13ba6ce3cc438fd064eecb5544

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95889 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15503 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5444 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100948 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46398 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97934 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15798 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23935 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73119 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30342 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98892 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11829 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86615 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53446 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11561 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4360 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45733 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81730 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4465 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102977 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22956 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16741 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82154 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23207 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82635 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81513 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20450 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26107 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3557 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16293 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22919 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28074 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22578 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26057 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24319 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeIncomingSourceGStreamer.cpp:
```diff
@@ -158,18 +157,73 @@ void RealtimeIncomingSourceGStreamer::handleDownstreamEvent(GstElement* sink, GR
+        // small bitrate updates downstream causes excessive overhead. Only send out tag updates if the bitrate tags
+        // have changed by at least 20% or if any other tag has changed.
+        const double bitrateChangeThreshold = 0.2;
+        const char* bitrateTags[] = { GST_TAG_BITRATE, GST_TAG_NOMINAL_BITRATE, GST_TAG_MAXIMUM_BITRATE, GST_TAG_MINIMUM_BITRATE };
```
Can span be used here?

cadubentzen replied:
I simplified the code which made this line unnecessary.

philn commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeIncomingSourceGStreamer.cpp:
```diff
@@ -158,18 +157,31 @@ void RealtimeIncomingSourceGStreamer::handleDownstreamEvent(GstElement* sink, GR
+    }
+    case GST_EVENT_TAG: {
+        // Prevent overhead at startup, when baseparse emits many tag events with small bitrate updates, by applying some backpressure.
+        constexpr auto tagUpdateTimout = 1_s;
```
missing e somewhere there ;)

cadubentzen replied:
:see_no_evil: thanks!

philn reviewed (approved):
LGTM but please fix the typo ;)

cadubentzen committed:
[GStreamer][WebRTC] Excessive tag events with bitrate update at startup
https://bugs.webkit.org/show_bug.cgi?id=290045

Reviewed by Philippe Normand.

Codec parsers like h264parse emit too many tag events at startup, while carrying
only small bitrate updates. As time passes, bitrate updates are sent
less frequently.

Reduce the overhead in the pipelines at startup by forwarding tag events only every
1 second in RealtimeIncomingSourceGStreamer.

Canonical link: https://commits.webkit.org/292399@main
